src/age_types.rs

```diff
@@ -119,6 +139,60 @@ where
     }
 }
 
+/// Represents path in graph. Used during process of path deserialization
+// Supporting "only" <V, E> generic types implies that we should use enums to represent these types.
```

Comment:
Maybe it's the alcohol or a good party on the New Year's Eve, but I cannot understand this comment. Can you provide a more descriptive information? Why it implies usage of the enums? What kind of enums?  Try explaining it like for a drunken man (it's me now ;) )

Reply:
Haha yes of course!

Let me try to explain.
With a query (like `MATCH p=()-->() return p`) you will get all vertices and edges in the db.
For example let's say we get "User -[WorksAt]-> Workplace" and "Food -[Contains]-> Ingredient".
Kinda strange set of data, but let's ignore that for now.

We can't fit User, Workplace, Cake and Ingredient in `V` unless we use an enum or something like `Box<dyn SomeTrait>`. Same goes for edges.

So to mitigate this we could do something like:
```rust
// We already have the models defined above.

#[derive(Deserialize, Serialize, PartialEq)]
enum VertexType {
    User(User),
    Workplace(Workplace),
    Food(Food),
    Ingredient(Ingredient),
}

#[derive(Deserialize, Serialize, PartialEq)]
enum EdgeType {
    WorksAt(WorksAt),
    Contains(Contains),
}

// Now let's use the enums
// `result` holds the result from querying "MATCH p=()-->() return p",
for row in result {
    let path: Path<VertexType, EdgeType> = row.get(0);
}
```

Hopefully this made more sense 😆 
I guess the comment probably shouldn't exist there, but maybe it would be helpful to users with some documentation wrt this example.

Reply:
There's another interesting situation that follows from this enum technique that should perhaps be documented as well.
If we have two enums variants with equal contents then serde deserializes to the first that matches.

```rust
#[derive(Deserialize, Serialize)]
struct A {
    id: usize
}

#[derive(Deserialize, Serialize)]
struct B {
    id: usize
}

#[derive(Deserialize, Serialize)]
enum VertexType {
    A(A)
    B(B)
}
```
The contents of a `B` could be deserialized into a `A` here, resulting in something like
`Vertex { id: 123456789, label: "B", properties: A(A { id: 1 }) }`

One way to deal with this is would be to deserialize to a `Vertex` first (or a serde json `Value`), then match on the label, and finally deserialize that into the correct model 🤷 

```rust
// Not tested, just an idea.
fn from_str(vertex_string: str) -> VertexType {
    // Using Vertex<String> here because we're only interested in the label
    match serde_json::from_str::<Vertex<String>>(vertex_string) {
        Ok(vertex) => {
            match vertex.label() {
                "A" => match serde_json::from_str::<Vertex<A>>(vertex_string) {
                    Ok(a) => Ok(VertexType::A(a)),
                    Err(e) => Err(e),
                },
                "B" => match serde_json::from_value::<B>(value) {
                    Ok(b) => Ok(VertexType::B(b)),
                    Err(e) => Err(e),
                },
            }
        }
    }
}
```
But that's kinda ugly and also it's annoying to require the users of the lib to do this.

Another option is to add a required trait to vertex and edge, like  `Vertex<T: GetLabel>`. This can be used with a  custom de/serializer to create the desired effect.
Quite a bit of work to get it right.

Now the final technique, which I think is best, is to inform the users that they can use serde tagging. Something like
```rust
#[derive(Deserialize, Serialize, PartialEq)]
#[serde(tag = "type")]
enum VertexType {
    A(A)
    B(B)
}
```
If I now do the following
```cypher
CREATE(b:B {id: 1, type:'B'})  // type:'B' is the important part
MATCH (b: B) WHERE b.id = 1 RETURN b
```
```rust
let test: Vertex<VertexType> = row.get(0);
println!("{:?}", test);
```
I get `Vertex { id: 844424930131969, label: "B", properties: B(B { id: 1 }) }`

I'm _pretty_ sure that this works in all cases 😆

Also it's important to remember that if you're using `AgType` to save models you must remember to include `type: "A"` in the query.
```rust
client.execute_cypher(
    'my_graph',
    "CREATE(a: A {id: $id, type:'A'})",  // <- here
    Some(AgType::<VertexType>(
        VertexType::A(
            A { id: 2 }
        )
    )
)
```
This again means that at insertion you have to know what _variant_  you're inserting. Perhaps the best way of keeping control of this is to use a trait so that you can do `let type: String = vertex_type.get_type()`
To achieve having access to the trait of the variant structs in the enum you could use [enum_dispatch](https://crates.io/crates/enum_dispatch).
```rust
#[derive(Deserialize, Serialize, PartialEq)]
#[serde(tag = "type")]
#[enum_dispatch]
enum VertexType {
...

#[enum_dispatch(VertexType)]
trait GetType {
    fn get_type(&self) -> String;
}

impl GetType for A {
    fn get_type(&self) -> String {
        "A"
    }
}
```

Sorry about the wall of text, as you can see I've thought about this quite a bit 😅 
Anyways! I think it could be valuable to document the best solution because I suspect it would be a recurring usecase for developers. I could create an issue from this if you'd like.

Reply:
I would love it <3 Don't want to lose it!

Reply:
Created an issue here #5 😊
It's a little better thought out than the text above.